WebView/app/src/androidTest/java/com/android/samples/webviewdemo/MainActivityTest.kt

```diff
@@ -0,0 +1,159 @@
+        mainActivityRule.getActivity()
+        onWebView()
+            .withElement(findElement(Locator.ID, "location"))
+            .perform(webClick()) // Similar to perform(click())
+            .withElement(findElement(Locator.ID, "SF"))
+            .perform(webClick()) // Similar to perform(click())
+            .withElement(findElement(Locator.ID, "title"))
+            .check(webMatches(getText(), containsString("San Francisco")))
+    }
+
+    // Test for checking createJsObject
+    @Test
+    fun jsObjectIsInjectedAndContainsPostMessage() {
+    mainActivityRule.getActivity()
+    onWebView()
+        .check(
+            webMatches(
+                script("return jsObject && jsObject.postMessage != null;", castOrDie(Boolean::class.javaObjectType)),
+                `is`(true)
+            )
+        )
+    }
+
+    @Test
+    fun valueInCallback_compareValueInput_returnsTrue(){
+        mainActivityRule.getActivity()
+
+        // Setup
+        val webView = WebView(context)
+        val jsObjName = "jsObject"
+        val allowedOriginRules = setOf<String>("https://example.com")
+
+        // Create HTML
+        val htmlPage = "<!DOCTYPE html><html><body>" + "    <script>" + "        myObject.postMessage('hello');" + "    </script>" + "</body></html>"
+
+        // Create JsObject
+        MainActivity.createJsObject(
+            webView,
+            jsObjName,
+            allowedOriginRules
+        ) { message -> MainActivity.invokeShareIntent(message) }
+
+        //Inject JsObject into Html
+        webView.loadData(htmlPage, "text/html", "UTF-8")
+
+
+        //Call js code to invoke callback (in script tag of htmlPage)
+
+        // evaluate what comes out -> it should be hello
+        // *Note: "response from callback" is a place holder here I am unsure what should be placed there
+        assertEquals("response from callback", "hello")
+
+    }
+
+    @Test
+    // Checks that postMessage runs on the UI thread
+    fun checkingThreadCallbackRunsOn() {
+        mainActivityRule.getActivity()
+
+        // Setup
+        val webView = WebView(context)
+        val jsObjName = "jsObject"
+        val allowedOriginRules = setOf<String>("https://example.com")
+
+        // Create HTML
+        val htmlPage =
+            "<!DOCTYPE html><html><body>" + "    <script>" + "        jsObject.postMessage('hello');" + "    </script>" + "</body></html>"
+
+
+        // Create JsObject
+        MainActivity.createJsObject(
+            webView,
+            jsObjName,
+            allowedOriginRules
+        ) { message -> MainActivity.invokeShareIntent(message) }
+
+
+        // Inject JsObject into Html by loading it in the webview
+        webView.loadData(htmlPage, "text/html", "UTF-8")
+
+
+        // Use coroutine to go onto UI thread here?
+        // Call js code to invoke callback (in script tag of htmlPage)
+
+
+        // check that method is running on the UI thread
+        assertTrue(isUiThread())
```

Comment:
We don't want this assert here because instrumentation tests run on the "test thread," not the UI thread. I wrote an [internal google doc](http://go/no-asserts-on-main-thread) which goes into more details about this.

As a first step, try moving this into createJsObject's callback. That will return correct results, but comes with the drawback that failing assertion crashes the process instead of showing up as a failed test (the google doc has more context on this). If you can get that working, then you can improve this further by using a `SettableFuture<Boolean>` to plumb the value across threads (see my previous comment for pointers).